File: Passbook/simplepush.php
#!/usr/bin/php
<?php
if ($argc < 2) {
?>
  Usage:
  <?php echo $argv[0]; ?> <token> [<token> <token>...]

  <token> are the device tokens to send the push notification to.

<?php
} 

if ($argc < 2) {
    exit(1);
}

echo 'There are ' . ($argc - 1) . ' tokens.' . PHP_EOL;

// Put your private key's passphrase here:
$passphrase = 'foo';

// Put your alert message here:
$message = 'Yo';

////////////////////////////////////////////////////////////////////////////////

function send_notification($deviceToken, $apns_stream) {
    // Create the payload body
    $body['aps'] = array(
    // 	'alert' => $message,
        'sound' => 'default'
        );

    // Encode the payload as JSON
    $payload = json_encode($body);

    // Build the binary notification
    $msg = chr(0) . pack('n', 32) . pack('H*', $deviceToken) . pack('n', strlen($payload)) . $payload;

    // Send it to the server
    $result = fwrite($apns_stream, $msg, strlen($msg));

    if (!$result)
        echo 'Message not delivered' . PHP_EOL;
    else
        echo 'Message successfully delivered' . PHP_EOL;
}

$ctx = stream_context_create();
stream_context_set_option($ctx, 'ssl', 'local_cert', 'apple_push_notification.pem');
// stream_context_set_option($ctx, 'ssl', 'passphrase', $passphrase);

// Open a connection to the APNS server
$fp = stream_socket_client(
	'ssl://gateway.push.apple.com:2195', $err,
	$errstr, 60, STREAM_CLIENT_CONNECT|STREAM_CLIENT_PERSISTENT, $ctx);

if (!$fp)
	exit("Failed to connect: $err $errstr" . PHP_EOL);

echo 'Connected to APNS' . PHP_EOL;

for ($i = 1; $i < $argc; $i++) {
    send_notification($argv[$i], $fp);
}

// Close the connection to the server
fclose($fp);

File: Passbook/simplepush.py
#! /usr/bin/env python

import socket, ssl, json, struct
import sys

# device token returned when the iPhone application
# registers to receive alerts
# It should be the argument to this script

deviceToken = sys.argv[1]

thePayLoad = {
     'aps': {
          'sound':'default',
          },
     }

# Certificate issued by apple and converted to .pem format with openSSL
# Per Apple's Push Notification Guide (end of chapter 3), first export the cert in p12 format
# openssl pkcs12 -in cert.p12 -out cert.pem -nodes 
#   when prompted "Enter Import Password:" hit return
#
theCertfile = 'apple_push_notification.pem'
# 
theHost = ('gateway.push.apple.com', 2195)

# 
data = json.dumps(thePayLoad)

# Clear out spaces in the device token and convert to hex
deviceToken = deviceToken.replace(' ','')
#byteToken = bytes.fromhex(deviceToken) # Python 3
byteToken = deviceToken.decode('hex') # Python 2

theFormat = '!BH32sH%ds' % len(data)
theNotification = struct.pack(theFormat, 0, 32, byteToken, len(data), data)

# Create our connection using the certfile saved locally
ssl_sock = ssl.wrap_socket(socket.socket(socket.AF_INET, socket.SOCK_STREAM), certfile = theCertfile)
ssl_sock.connect(theHost)

# Write out our data
ssl_sock.write(theNotification)

# Close the connection -- apple would prefer that we keep
# a connection open and push data as needed.
ssl_sock.close()

print "notification sent for " + deviceToken
